mkosi/gentoo.py

```diff
@@ -0,0 +1,482 @@
+            self.pkgs_fs.append("cryptsetup")
+            self.pkgs_fs.append("device-mapper")
+
+        self.grub_platforms = list()
+        if not do_run_build_script and args.bootable:
+            if args.esp_partno:
+                self.pkgs_boot = ["sys-kernel/installkernel-systemd-boot"]
+            elif args.bios_partno:
+                self.pkgs_boot = ["sys-boot/grub"]
+                self.grub_platforms = ["coreboot", "qemu", "pc"]
+
+            self.pkgs_boot.append("sys-kernel/gentoo-kernel-bin")
+            self.pkgs_boot.append("sys-firmware/edk2-ovmf")
+
+        self.UNINSTALL_IGNORE = [
+            "/bin",
+            "/sbin",
+            "/lib",
+            "/lib64",
+        ]
+
+        # GENTOO_DONTMOVE: self.grub_platforms, for instance, must be set
+        self.emerge_vars = {
+            "BOOTSTRAP_USE": " ".join(self.portage_use_flags),
+            "FEATURES": " ".join(self.portage_features),
+            "EGIT_CLONE_TYPE": "shallow",
+            "GRUB_PLATFORMS": " ".join(self.grub_platforms),
+            "UNINSTALL_IGNORE": " ".join(self.UNINSTALL_IGNORE),
+            "USE": " ".join(self.portage_use_flags),
+        }
+
+        self.fetch_fix_stage3(root)
+        self.set_profile(args, root)
+        self.set_default_repo()
+        self.unmask_arch()
+        self.whitelist_licenses()
+        self.provide_patches()
+        self.set_useflags(root)
+        self.mkosi_conf(root)
+        self.invoke_emerge(args, root, inside_stage3=False, actions=["--sync"])
+        self.baselayout(args, root)
+        self.update_stage3(args, root)
+
+        if "build-script" in ARG_DEBUG:
+            self.invoke_emerge(args, root, actions=["--info"])
+
+    def fetch_fix_stage3(self, root: Path) -> None:
+        """usrmerge tracker bug: https://bugs.gentoo.org/690294"""
+
+        # e.g.: http://distfiles.gentoo.org/releases/amd64/autobuilds/latest-stage3.txt
+        stage3tsf_path_url = urllib.parse.urljoin(
+            self.portage_cfg["GENTOO_MIRRORS"].partition(" ")[0],
+            f"releases/{self.arch}/autobuilds/latest-stage3.txt",
+        )
+        stage3_tar_str = ""
+        # GENTOO_UPSTREAM: wait for fix upstream: https://bugs.gentoo.org/792081
+        # and more... so we can gladly escape all this hideousness!
+        with urllib.request.urlopen(stage3tsf_path_url) as r:
+            args_profile = "nomultilib"
+            # 20210323T005051Z/stage3-arm64-systemd-20210323T005051Z.tar.xz 196362344
+            # 20210711T170538Z/stage3-amd64-nomultilib-systemd-20210711T170538Z.tar.xz 214470580
+            regexp = f"^[0-9TZ]+[/]stage3-{self.arch}-{args_profile}-systemd-[0-9TZ]+[.]tar[.]xz"
+            all_lines = r.readlines()
+            for line in all_lines:
+                m = re.match(regexp, line.decode("utf-8"))
+                if m:
+                    stage3_tar_str = m.group(0)
+
+            if not stage3_tar_str:
+                die("profile names changed upstream?")
```

Comment:
```py
            for line in all_lines:
                m = re.match(regexp, line.decode("utf-8"))
                if m:
                    stage3_tar = Path(m.group(0))
                    break
            else:
                die("profile names changed upstream?")
```

Reply:
not sure about this one. on the first `line` we `break` out of the `for` loop `if` it matches `else`, i.e. `all_lines[0]` doesn't match, we `die`?

Reply:
Yes. The difference with your code is that the *first* match wins, not the last. If that's wrong, and the last should win, then remove the break.

Reply:
this will `die` on the very first line if the first line doesn't match which doesn't.

Reply:
No, the `else` is attached to `for`, not the conditional.

Reply:
nice :)